0000-access-keys.md

```diff
@@ -0,0 +1,286 @@
+
+An example of an access key limited to `proxy` function:
+```rust
+/// vasya.near,3bc2b3b
+AccessKey {
+    nonce: 0,
+    
+    allowance: 1_000_000_000,
+    
+    receiver_id: "vasya.near",
+    
+    method_name: Some("proxy"),
+
+    full_access: false,
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+- Access keys are going to be stored with the `account_id` prefix and the `public_key` suffix.
+They should be stored on the same shard as the account.
+- Access keys storage rent should be accounted and paid from the account directly without affecting the allowance.
+- Access keys allowance can exceed the account balance.
+- To validate a transaction signed with the AccessKey, we need to first validate the signature, then fetch the Account and the AccessKey, validate that we have enough funds and verify permissions. 
+- Account creation should now create a full permission access keys with MAX allowance, instead of public keys within the account.
+- SwapKey transaction should just replace the old access key with the given new access key.
+
+### Technical changes
+
+#### `nonce` on the AccessKey level instead of account level
+
+Since access keys can be used by the different people or parties at the same time, we need to be able to 
+have separate nonce for each key instead of a single nonce at the account level.
+With a single nonce on the account level, there is a high probability that 2 apps would use the same nonce for 2 different transactions and one of this transactions would be rejected.
+
+Previously we were ordering transactions by nonce and rejecting transactions with a duplicated or lower nonce.
+With the access key nonce, we still need to order transactions by nonce, but now we need to group them by `account_id,public_key` key instead of just account_id.
+
+#### `allowance` field
+
+Allowance is the amount of tokens the AccessKey can spend from the account balance.
+When some amount is spent, it's subtracted from both the allowance of the access key and from the account balance.
+
+NOTE: In the previous iteration of access keys, we used balance instead of the allowance. But it required to sum up all access keys balances to get the the total account balance.
+It also prevented sharing of the account balance between access keys.
+
+#### Permissions
+
+Almost all desired use-cases of access keys can be achieved by using the old permissions model.
+It restricts allows access keys to only issue function call with no attached tokens.
+The function calls are restricted to the selected `receiver_id` and potentially restricted to a single `method_name`.
+Anything non-trivial can be done by the contract that receives this call, e.g. through `proxy` function.
+
+If we want to remove public keys from the account, we also need to add a new permission that allows unlimited access to the account.
```

Comment:
Unless a contract is locked, why would there not be a full access key?

Reply:
By default access keys allow only function calls

Reply:
replaced permissions model, to make it clear